File: ios/QingKeSchedule/Features/SchedulePresentation.swift
import Foundation

struct TodayCourseItem: Equatable, Identifiable {
    let occurrence: CourseOccurrenceDTO
    let status: CourseStatus
    let isNext: Bool
    let timingProgress: CourseTimingProgress?

    var id: String { occurrence.schedule.id }
}

struct CourseTimingProgress: Equatable {
    let elapsedSeconds: Int
    let remainingSeconds: Int
    let fraction: Double

    var elapsedMinutes: Int { elapsedSeconds / 60 }

    var remainingClockText: String {
        String(format: "%d:%02d", remainingSeconds / 60, remainingSeconds % 60)
    }
}

struct TodaySchedulePresentation: Equatable {
    let teachingWeek: Int?
    let items: [TodayCourseItem]
    let emptyMessage: String
    let isNonTeachingDay: Bool

    init(
        semester: SemesterDTO,
        courses: [CourseDTO],
        now: Date,
        academicCalendarSettings: AcademicCalendarSettings = .defaults,
        calendar: Calendar
    ) {
        let calculatedWeek = ScheduleRules.teachingWeek(
            for: now,
            semester: semester,
            calendar: calendar
        )
        teachingWeek = calculatedWeek

        let resolution = academicCalendarSettings.resolution(for: now, calendar: calendar)
        let occurrences: [CourseOccurrenceDTO]
        switch resolution {
        case .nonTeaching:
            isNonTeachingDay = true
            occurrences = []
        case .teaching(let sourceDayOfWeek, _):
            isNonTeachingDay = false
            if let calculatedWeek,
               ScheduleRules.isTeachingWeekInSemester(calculatedWeek, semester: semester) {
                occurrences = ScheduleRules.occurrences(
                    forWeek: calculatedWeek,
                    courses: courses
                )
                .filter { $0.schedule.dayOfWeek == sourceDayOfWeek }
                .enumerated()
                .sorted(by: Self.stableOccurrenceOrder)
                .map(\.element)
            } else {
                occurrences = []
            }
        }
        let statuses = occurrences.map {
            ScheduleRules.occurrenceStatus(
                $0,
                semester: semester,
                now: now,
                calendar: calendar
            )
        }
        let nextIndex = statuses.firstIndex(of: .upcoming)
        items = occurrences.enumerated().map { index, occurrence in
            TodayCourseItem(
                occurrence: occurrence,
                status: statuses[index],
                isNext: index == nextIndex,
                timingProgress: Self.timingProgress(
                    occurrence: occurrence,
                    status: statuses[index],
                    semester: semester,
                    now: now,
                    calendar: calendar
                )
            )
        }

        if case .nonTeaching(let reason) = resolution,
           let calculatedWeek,
           ScheduleRules.isTeachingWeekInSemester(calculatedWeek, semester: semester) {
            emptyMessage = "\(reason)，今日不显示课程。"
        } else if let calculatedWeek,
           ScheduleRules.isTeachingWeekInSemester(calculatedWeek, semester: semester) {
            emptyMessage = "今天没有课程，享受空闲时间吧。"
        } else {
            emptyMessage = "当前日期不在这个学期内。"
        }
    }

    private static func stableOccurrenceOrder(
        _ left: (offset: Int, element: CourseOccurrenceDTO),
        _ right: (offset: Int, element: CourseOccurrenceDTO)
    ) -> Bool {
        if left.element.schedule.startPeriod != right.element.schedule.startPeriod {
            return left.element.schedule.startPeriod < right.element.schedule.startPeriod
        }
        if left.element.schedule.endPeriod != right.element.schedule.endPeriod {
            return left.element.schedule.endPeriod < right.element.schedule.endPeriod
        }
        let nameOrder = left.element.course.name.compare(
            right.element.course.name,
            locale: Locale(identifier: "zh_CN")
        )
        if nameOrder != .orderedSame { return nameOrder == .orderedAscending }
        return left.offset < right.offset
    }

    private static func timingProgress(
        occurrence: CourseOccurrenceDTO,
        status: CourseStatus,
        semester: SemesterDTO,
        now: Date,
        calendar: Calendar
    ) -> CourseTimingProgress? {
        guard status == .ongoing,
              let startPeriod = semester.periods.first(where: {
                  $0.number == occurrence.schedule.startPeriod
              }),
              let endPeriod = semester.periods.first(where: {
                  $0.number == occurrence.schedule.endPeriod
              }),
              let startMinutes = ScheduleRules.minutes(from: startPeriod.startTime),
              let endMinutes = ScheduleRules.minutes(from: endPeriod.endTime)
        else {
            return nil
        }

        let startOfDay = calendar.startOfDay(for: now)
        guard let startDate = calendar.date(
            byAdding: .second,
            value: startMinutes * 60,
            to: startOfDay
        ), var endDate = calendar.date(
            byAdding: .second,
            value: endMinutes * 60,
            to: startOfDay
        ) else {
            return nil
        }
        if endDate <= startDate {
            guard let followingDayEnd = calendar.date(byAdding: .day, value: 1, to: endDate) else {
                return nil
            }
            endDate = followingDayEnd
        }

        let duration = max(endDate.timeIntervalSince(startDate), 1)
        let elapsed = min(max(now.timeIntervalSince(startDate), 0), duration)
        let remaining = max(endDate.timeIntervalSince(now), 0)
        return CourseTimingProgress(
            elapsedSeconds: Int(elapsed.rounded(.down)),
            remainingSeconds: Int(remaining.rounded(.up)),
            fraction: Double(elapsed) / Double(duration)
        )
    }
}

struct WeekCourseItem: Equatable, Identifiable {
    let occurrence: CourseOccurrenceDTO
    let isConflicting: Bool
    let displayDayOfWeek: Int

    var id: String { "\(displayDayOfWeek)-\(occurrence.schedule.id)" }
}

struct WeekDayPresentation: Equatable, Identifiable {
    let dayOfWeek: Int
    let date: Date?
    let items: [WeekCourseItem]
    let isNonTeachingDay: Bool
    let scheduleSourceDayOfWeek: Int?

    var id: Int { dayOfWeek }
}

struct WeekMatrixItem: Equatable, Identifiable {
    let id: String
    let occurrence: CourseOccurrenceDTO
    let isConflicting: Bool
    let dayColumn: Int
    let startRow: Int
    let rowSpan: Int
    let lane: Int
    let laneCount: Int

}

struct WeekMatrixBreak: Equatable {
    let title: String
    let startTime: String
    let endTime: String
    let insertionRow: Int
}

struct WeekMatrixPresentation: Equatable {
    let periods: [PeriodDTO]
    let items: [WeekMatrixItem]
    let scheduleBreak: WeekMatrixBreak?

    init(
        semester: SemesterDTO,
        days: [WeekDayPresentation],
        academicCalendarSettings: AcademicCalendarSettings = .defaults
    ) {
        periods = semester.periods.sorted { $0.number < $1.number }
        scheduleBreak = Self.makeBreak(
            academicCalendarSettings.lunchBreak,
            periods: periods
        )
        let periodRows = Dictionary(uniqueKeysWithValues: periods.enumerated().map {
            ($0.element.number, $0.offset)
        })

        items = days
            .filter { (1...ScheduleDisplayText.weekdayNames.count).contains($0.dayOfWeek) }
            .flatMap { day in
                let drafts = day.items.compactMap { item -> Draft? in
                    guard
                        let startRow = periodRows[item.occurrence.schedule.startPeriod],
                        let endRow = periodRows[item.occurrence.schedule.endPeriod],
                        endRow >= startRow
                    else {
                        return nil
                    }
                    return Draft(
                        item: item,
                        dayColumn: day.dayOfWeek - 1,
                        startRow: startRow,
                        rowSpan: endRow - startRow + 1
                    )
                }
                return Self.laidOut(drafts)
            }
            .sorted {
                if $0.dayColumn != $1.dayColumn { return $0.dayColumn < $1.dayColumn }
                if $0.startRow != $1.startRow { return $0.startRow < $1.startRow }
                if $0.lane != $1.lane { return $0.lane < $1.lane }
                return $0.id < $1.id
            }
    }

    private static func makeBreak(
        _ settings: ScheduleBreakSettings,
        periods: [PeriodDTO]
    ) -> WeekMatrixBreak? {
        guard
            settings.isEnabled,
            let breakStart = ScheduleRules.minutes(from: settings.startTime),
            let breakEnd = ScheduleRules.minutes(from: settings.endTime),
            let insertionRow = periods.firstIndex(where: {
                guard let start = ScheduleRules.minutes(from: $0.startTime) else { return false }
                return start >= breakEnd
            }),
            insertionRow > 0,
            let previousEnd = ScheduleRules.minutes(from: periods[insertionRow - 1].endTime),
            previousEnd <= breakStart
        else {
            return nil
        }
        return WeekMatrixBreak(
            title: settings.title,
            startTime: settings.startTime,
            endTime: settings.endTime,
            insertionRow: insertionRow
        )
    }

    private struct Draft {
        let item: WeekCourseItem
        let dayColumn: Int
        let startRow: Int
        let rowSpan: Int

        var endRow: Int { startRow + rowSpan - 1 }
    }

    private static func laidOut(_ drafts: [Draft]) -> [WeekMatrixItem] {
        let sorted = drafts.sorted {
            if $0.startRow != $1.startRow { return $0.startRow < $1.startRow }
            if $0.endRow != $1.endRow { return $0.endRow < $1.endRow }
            return $0.item.id < $1.item.id
        }
        var result: [WeekMatrixItem] = []
        var component: [Draft] = []
        var componentEnd = -1

        for draft in sorted {
            if !component.isEmpty, draft.startRow > componentEnd {
                result.append(contentsOf: layoutComponent(component))
                component.removeAll(keepingCapacity: true)
                componentEnd = -1
            }
            component.append(draft)
            componentEnd = max(componentEnd, draft.endRow)
        }
        result.append(contentsOf: layoutComponent(component))
        return result
    }

    private static func layoutComponent(_ drafts: [Draft]) -> [WeekMatrixItem] {
        guard !drafts.isEmpty else { return [] }
        var laneEndRows: [Int] = []
        var placements: [(draft: Draft, lane: Int)] = []

        for draft in drafts {
            let lane: Int
            if let reusableLane = laneEndRows.firstIndex(where: { $0 < draft.startRow }) {
                lane = reusableLane
                laneEndRows[lane] = draft.endRow
            } else {
                lane = laneEndRows.count
                laneEndRows.append(draft.endRow)
            }
            placements.append((draft, lane))
        }

        let laneCount = laneEndRows.count
        return placements.map { placement in
            WeekMatrixItem(
                id: placement.draft.item.id,
                occurrence: placement.draft.item.occurrence,
                isConflicting: placement.draft.item.isConflicting,
                dayColumn: placement.draft.dayColumn,
                startRow: placement.draft.startRow,
                rowSpan: placement.draft.rowSpan,
                lane: placement.lane,
                laneCount: laneCount
            )
        }
    }
}

struct WeekSchedulePresentation: Equatable {
    let week: Int
    let currentWeek: Int?
    let days: [WeekDayPresentation]

    init(
        week: Int,
        semester: SemesterDTO,
        courses: [CourseDTO],
        now: Date,
        academicCalendarSettings: AcademicCalendarSettings = .defaults,
        calendar: Calendar
    ) {
        let resolvedWeek = min(max(week, 1), semester.totalWeeks)
        self.week = resolvedWeek
        let calculatedCurrentWeek = ScheduleRules.teachingWeek(
            for: now,
            semester: semester,
            calendar: calendar
        )
        currentWeek = calculatedCurrentWeek.flatMap {
            ScheduleRules.isTeachingWeekInSemester($0, semester: semester) ? $0 : nil
        }

        let occurrences = ScheduleRules.occurrences(forWeek: resolvedWeek, courses: courses)
        days = (1...ScheduleDisplayText.weekdayNames.count).map { dayOfWeek in
            let date = ScheduleRules.date(
                forTeachingWeek: resolvedWeek,
                dayOfWeek: dayOfWeek,
                semester: semester,
                calendar: calendar
            )
            let resolution = date.map {
                academicCalendarSettings.resolution(for: $0, calendar: calendar)
            }
            let sourceDayOfWeek: Int?
            let isNonTeachingDay: Bool
            switch resolution {
            case .teaching(let source, _):
                sourceDayOfWeek = source
                isNonTeachingDay = false
            case .nonTeaching, .none:
                sourceDayOfWeek = nil
                isNonTeachingDay = resolution != nil
            }
            let dayOccurrences = sourceDayOfWeek.map { source in
                occurrences
                    .filter { $0.schedule.dayOfWeek == source }
                    .sorted(by: Self.stableOccurrenceOrder)
            } ?? []
            let conflictIDs = Self.conflictingScheduleIDs(
                in: dayOccurrences,
                week: resolvedWeek
            )
            return WeekDayPresentation(
                dayOfWeek: dayOfWeek,
                date: date,
                items: dayOccurrences.map {
                    WeekCourseItem(
                        occurrence: $0,
                        isConflicting: conflictIDs.contains($0.schedule.id),
                        displayDayOfWeek: dayOfWeek
                    )
                },
                isNonTeachingDay: isNonTeachingDay,
                scheduleSourceDayOfWeek: sourceDayOfWeek
            )
        }
    }

    static func initialWeek(
        semester: SemesterDTO,
        now: Date,
        calendar: Calendar
    ) -> Int {
        guard let current = ScheduleRules.teachingWeek(
            for: now,
            semester: semester,
            calendar: calendar
        ) else {
            return 1
        }
        return min(max(current, 1), semester.totalWeeks)
    }

    private static func conflictingScheduleIDs(
        in occurrences: [CourseOccurrenceDTO],
        week: Int
    ) -> Set<String> {
        var identifiers = Set<String>()
        for leftIndex in occurrences.indices {
            for rightIndex in occurrences.indices where rightIndex > leftIndex {
                let left = occurrences[leftIndex]
                let right = occurrences[rightIndex]
                guard left.course.id != right.course.id,
                      ScheduleRules.scheduleApplies(left.schedule, inWeek: week),
                      ScheduleRules.scheduleApplies(right.schedule, inWeek: week),
                      left.schedule.dayOfWeek == right.schedule.dayOfWeek,
                      ScheduleRules.periodRangesOverlap(left.schedule, right.schedule)
                else {
                    continue
                }
                identifiers.insert(left.schedule.id)
                identifiers.insert(right.schedule.id)
            }
        }
        return identifiers
    }

    private static func stableOccurrenceOrder(
        _ left: CourseOccurrenceDTO,
        _ right: CourseOccurrenceDTO
    ) -> Bool {
        if left.schedule.startPeriod != right.schedule.startPeriod {
            return left.schedule.startPeriod < right.schedule.startPeriod
        }
        if left.schedule.endPeriod != right.schedule.endPeriod {
            return left.schedule.endPeriod < right.schedule.endPeriod
        }
        if left.course.name != right.course.name {
            return left.course.name.localizedCompare(right.course.name) == .orderedAscending
        }
        return left.schedule.id < right.schedule.id
    }
}

File: Android/app/src/main/java/com/qingke/schedule/presentation/SchedulePresentation.kt
package com.qingke.schedule.presentation

import com.qingke.schedule.domain.Course
import com.qingke.schedule.domain.CourseOccurrence
import com.qingke.schedule.domain.CourseSchedule
import com.qingke.schedule.domain.CourseStatus
import com.qingke.schedule.domain.OccurrenceKey
import com.qingke.schedule.domain.Period
import com.qingke.schedule.domain.ScheduleRules
import com.qingke.schedule.domain.Semester
import com.qingke.schedule.preferences.AcademicCalendarPreferences
import com.qingke.schedule.preferences.LunchBreakSettings
import java.text.Collator
import java.time.LocalDate
import java.time.LocalDateTime
import java.util.Locale

sealed interface AcademicDayResolution {
    data class TeachingDay(val sourceDayOfWeek: Int, val isMakeup: Boolean) : AcademicDayResolution
    data class NonTeachingDay(val reason: String) : AcademicDayResolution
}

object AcademicCalendarResolver {
    fun resolve(date: LocalDate, preferences: AcademicCalendarPreferences): AcademicDayResolution {
        val value = date.toString()
        if (value in preferences.nonTeachingDates) {
            return AcademicDayResolution.NonTeachingDay("已设为停课日")
        }
        preferences.makeupTeachingDays.firstOrNull { it.date == value }?.let { makeup ->
            return AcademicDayResolution.TeachingDay(makeup.followsDayOfWeek, isMakeup = true)
        }
        val dayOfWeek = date.dayOfWeek.value
        if (preferences.weekendsAreNonTeachingDays && dayOfWeek >= 6) {
            return AcademicDayResolution.NonTeachingDay("周末默认停课")
        }
        return AcademicDayResolution.TeachingDay(dayOfWeek, isMakeup = false)
    }
}

data class CourseTimingProgress(
    val elapsedMinutes: Int,
    val remainingMinutes: Int,
    val fraction: Double,
)

data class TodayCourseItem(
    val occurrence: CourseOccurrence,
    val status: CourseStatus,
    val isNext: Boolean,
    val timingProgress: CourseTimingProgress?,
)

data class TodaySchedulePresentation(
    val teachingWeek: Int?,
    val items: List<TodayCourseItem>,
    val emptyMessage: String,
    val isNonTeachingDay: Boolean,
) {
    companion object {
        fun create(
            semester: Semester,
            courses: List<Course>,
            now: LocalDateTime,
            academicCalendar: AcademicCalendarPreferences = AcademicCalendarPreferences.defaults,
        ): TodaySchedulePresentation {
            val week = ScheduleRules.teachingWeek(now.toLocalDate(), semester)
            val resolution = AcademicCalendarResolver.resolve(now.toLocalDate(), academicCalendar)
            val occurrences = when {
                resolution !is AcademicDayResolution.TeachingDay -> emptyList()
                !ScheduleRules.isTeachingWeekInSemester(week, semester) -> emptyList()
                else -> ScheduleRules.occurrencesForWeek(week, courses)
                    .filter { it.schedule.dayOfWeek == resolution.sourceDayOfWeek }
                    .sortedWith(occurrenceComparator)
            }
            val statuses = occurrences.map { ScheduleRules.occurrenceStatus(it, semester, now) }
            val nextIndex = statuses.indexOfFirst { it == CourseStatus.UPCOMING }
            val items = occurrences.mapIndexed { index, occurrence ->
                val status = statuses[index]
                TodayCourseItem(
                    occurrence = occurrence,
                    status = status,
                    isNext = index == nextIndex,
                    timingProgress = timingProgress(occurrence, status, semester, now),
                )
            }
            val inSemester = ScheduleRules.isTeachingWeekInSemester(week, semester)
            val message = when {
                resolution is AcademicDayResolution.NonTeachingDay && inSemester ->
                    "${resolution.reason}，今日不显示课程。"
                inSemester -> "今天没有课程，享受空闲时间吧。"
                else -> "当前日期不在这个学期内。"
            }
            return TodaySchedulePresentation(
                teachingWeek = week,
                items = items,
                emptyMessage = message,
                isNonTeachingDay = resolution is AcademicDayResolution.NonTeachingDay,
            )
        }
    }
}

data class WeekCourseItem(
    val occurrence: CourseOccurrence,
    val isConflicting: Boolean,
    val displayDayOfWeek: Int,
)

data class WeekDayPresentation(
    val dayOfWeek: Int,
    val date: LocalDate,
    val items: List<WeekCourseItem>,
    val isNonTeachingDay: Boolean,
    val scheduleSourceDayOfWeek: Int?,
)

data class WeekSchedulePresentation(
    val week: Int,
    val currentWeek: Int?,
    val days: List<WeekDayPresentation>,
) {
    companion object {
        fun create(
            requestedWeek: Int,
            semester: Semester,
            courses: List<Course>,
            now: LocalDateTime,
            academicCalendar: AcademicCalendarPreferences = AcademicCalendarPreferences.defaults,
        ): WeekSchedulePresentation {
            val resolvedWeek = requestedWeek.coerceIn(1, semester.totalWeeks)
            val calculatedCurrentWeek = ScheduleRules.teachingWeek(now.toLocalDate(), semester)
            val currentWeek = calculatedCurrentWeek.takeIf { ScheduleRules.isTeachingWeekInSemester(it, semester) }
            val occurrences = ScheduleRules.occurrencesForWeek(resolvedWeek, courses)
            val days = (1..7).map { displayDay ->
                val date = requireNotNull(ScheduleRules.dateForTeachingWeek(resolvedWeek, displayDay, semester))
                when (val resolution = AcademicCalendarResolver.resolve(date, academicCalendar)) {
                    is AcademicDayResolution.NonTeachingDay -> WeekDayPresentation(
                        dayOfWeek = displayDay,
                        date = date,
                        items = emptyList(),
                        isNonTeachingDay = true,
                        scheduleSourceDayOfWeek = null,
                    )
                    is AcademicDayResolution.TeachingDay -> {
                        val items = occurrences
                            .filter { it.schedule.dayOfWeek == resolution.sourceDayOfWeek }
                            .sortedWith(occurrenceComparator)
                        val conflicts = conflictingKeys(items)
                        WeekDayPresentation(
                            dayOfWeek = displayDay,
                            date = date,
                            items = items.map { WeekCourseItem(it, it.key in conflicts, displayDay) },
                            isNonTeachingDay = false,
                            scheduleSourceDayOfWeek = resolution.sourceDayOfWeek,
                        )
                    }
                }
            }
            return WeekSchedulePresentation(resolvedWeek, currentWeek, days)
        }

        fun initialWeek(semester: Semester, now: LocalDateTime): Int =
            ScheduleRules.teachingWeek(now.toLocalDate(), semester).coerceIn(1, semester.totalWeeks)
    }
}

data class WeekMatrixItem(
    val id: String,
    val occurrence: CourseOccurrence,
    val isConflicting: Boolean,
    val dayColumn: Int,
    val startRow: Int,
    val rowSpan: Int,
    val lane: Int,
    val laneCount: Int,
)

data class WeekMatrixBreak(
    val title: String,
    val startTime: String,
    val endTime: String,
    val insertionRow: Int,
)

data class WeekMatrixPresentation(
    val periods: List<Period>,
    val items: List<WeekMatrixItem>,
    val scheduleBreak: WeekMatrixBreak?,
) {
    companion object {
        fun create(
            semester: Semester,
            days: List<WeekDayPresentation>,
            academicCalendar: AcademicCalendarPreferences = AcademicCalendarPreferences.defaults,
        ): WeekMatrixPresentation {
            val periods = semester.periods.sortedBy { it.number }
            val periodRows = periods.mapIndexed { index, period -> period.number to index }.toMap()
            val items = days.filter { it.dayOfWeek in 1..7 }.flatMap { day ->
                layoutDay(day, periodRows)
            }.sortedWith(compareBy<WeekMatrixItem> { it.dayColumn }
                .thenBy { it.startRow }
                .thenBy { it.lane }
                .thenBy { it.occurrence.key.courseIndex }
                .thenBy { it.occurrence.key.scheduleIndex })
            return WeekMatrixPresentation(periods, items, makeBreak(academicCalendar.lunchBreak, periods))
        }

        private fun layoutDay(day: WeekDayPresentation, periodRows: Map<Int, Int>): List<WeekMatrixItem> {
            val drafts = day.items.mapNotNull { item ->
                val startRow = periodRows[item.occurrence.schedule.startPeriod] ?: return@mapNotNull null
                val endRow = periodRows[item.occurrence.schedule.endPeriod] ?: return@mapNotNull null
                if (endRow < startRow) return@mapNotNull null
                MatrixDraft(item, day.dayOfWeek - 1, startRow, endRow - startRow + 1)
            }.sortedWith(compareBy<MatrixDraft> { it.startRow }
                .thenBy { it.endRow }
                .thenBy { it.item.occurrence.key.courseIndex }
                .thenBy { it.item.occurrence.key.scheduleIndex })
            val result = mutableListOf<WeekMatrixItem>()
            val component = mutableListOf<MatrixDraft>()
            var componentEnd = -1
            drafts.forEach { draft ->
                if (component.isNotEmpty() && draft.startRow > componentEnd) {
                    result += layoutComponent(component)
                    component.clear()
                    componentEnd = -1
                }
                component += draft
                componentEnd = maxOf(componentEnd, draft.endRow)
            }
            result += layoutComponent(component)
            return result
        }

        private fun layoutComponent(component: List<MatrixDraft>): List<WeekMatrixItem> {
            if (component.isEmpty()) return emptyList()
            val laneEnds = mutableListOf<Int>()
            val placements = component.map { draft ->
                val lane = laneEnds.indexOfFirst { it < draft.startRow }.takeIf { it >= 0 } ?: laneEnds.size
                if (lane == laneEnds.size) laneEnds += draft.endRow else laneEnds[lane] = draft.endRow
                draft to lane
            }
            return placements.map { (draft, lane) ->
                WeekMatrixItem(
                    id = "${draft.dayColumn}:${draft.item.occurrence.key.courseIndex}:${draft.item.occurrence.key.scheduleIndex}",
                    occurrence = draft.item.occurrence,
                    isConflicting = draft.item.isConflicting,
                    dayColumn = draft.dayColumn,
                    startRow = draft.startRow,
                    rowSpan = draft.rowSpan,
                    lane = lane,
                    laneCount = laneEnds.size,
                )
            }
        }

        private fun makeBreak(settings: LunchBreakSettings, periods: List<Period>): WeekMatrixBreak? {
            if (!settings.isEnabled) return null
            val start = ScheduleRules.minutes(settings.startTime) ?: return null
            val end = ScheduleRules.minutes(settings.endTime) ?: return null
            if (start >= end) return null
            val insertion = periods.indexOfFirst { period ->
                (ScheduleRules.minutes(period.startTime) ?: Int.MIN_VALUE) >= end
            }
            if (insertion <= 0) return null
            val previousEnd = ScheduleRules.minutes(periods[insertion - 1].endTime) ?: return null
            if (previousEnd > start) return null
            return WeekMatrixBreak(settings.title, settings.startTime, settings.endTime, insertion)
        }
    }

    private data class MatrixDraft(
        val item: WeekCourseItem,
        val dayColumn: Int,
        val startRow: Int,
        val rowSpan: Int,
    ) {
        val endRow: Int get() = startRow + rowSpan - 1
    }
}

object ScheduleDisplayText {
    val weekdayNames = listOf("周一", "周二", "周三", "周四", "周五", "周六", "周日")

    fun weekdayName(dayOfWeek: Int): String = weekdayNames.getOrElse(dayOfWeek - 1) { "" }

    fun periodRange(schedule: CourseSchedule): String = if (schedule.startPeriod == schedule.endPeriod) {
        "第 ${schedule.startPeriod} 节"
    } else {
        "第 ${schedule.startPeriod}–${schedule.endPeriod} 节"
    }

    fun timeRange(schedule: CourseSchedule, semester: Semester): String {
        val start = semester.periods.firstOrNull { it.number == schedule.startPeriod }?.startTime ?: "--:--"
        val end = semester.periods.firstOrNull { it.number == schedule.endPeriod }?.endTime ?: "--:--"
        return "$start–$end"
    }

    fun weekMatrixSummary(periodCount: Int): String = "MON–SUN / $periodCount PERIODS"

    fun compactCourseDetails(course: Course, schedule: CourseSchedule): String =
        listOf(schedule.classroom, course.teacher).filter { it.isNotEmpty() }.joinToString(" · ")
}

private val occurrenceComparator = Comparator<CourseOccurrence> { left, right ->
    compareValuesBy(left, right, { it.schedule.startPeriod }, { it.schedule.endPeriod }).takeIf { it != 0 }
        ?: zhCollator.compare(left.course.name, right.course.name).takeIf { it != 0 }
        ?: compareValuesBy(left, right, { it.key.courseIndex }, { it.key.scheduleIndex })
}

private val zhCollator: Collator = Collator.getInstance(Locale.SIMPLIFIED_CHINESE)

private fun timingProgress(
    occurrence: CourseOccurrence,
    status: CourseStatus,
    semester: Semester,
    now: LocalDateTime,
): CourseTimingProgress? {
    if (status != CourseStatus.ONGOING) return null
    val start = semester.periods.firstOrNull { it.number == occurrence.schedule.startPeriod }
        ?.let { ScheduleRules.minutes(it.startTime) } ?: return null
    val end = semester.periods.firstOrNull { it.number == occurrence.schedule.endPeriod }
        ?.let { ScheduleRules.minutes(it.endTime) } ?: return null
    val current = now.hour * 60 + now.minute
    val duration = maxOf(end - start, 1)
    val elapsed = (current - start).coerceIn(0, duration)
    return CourseTimingProgress(elapsed, maxOf(end - current, 0), elapsed.toDouble() / duration)
}

private fun conflictingKeys(occurrences: List<CourseOccurrence>): Set<OccurrenceKey> {
    val conflicts = mutableSetOf<OccurrenceKey>()
    occurrences.indices.forEach { leftIndex ->
        (leftIndex + 1 until occurrences.size).forEach { rightIndex ->
            val left = occurrences[leftIndex]
            val right = occurrences[rightIndex]
            if (left.course.id != right.course.id &&
                ScheduleRules.periodRangesOverlap(left.schedule, right.schedule)
            ) {
                conflicts += left.key
                conflicts += right.key
            }
        }
    }
    return conflicts
}
